WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingViewModel.kt

```diff
@@ -1,13 +1,199 @@
+        }
+        // TODO all the other screens that can be persisted.
+    }
+
+    private fun navigateToScreen(prepublishingScreen: PrepublishingScreen) {
+        when (prepublishingScreen) {
+            HOME -> updateNavigationTarget(PrepublishingNavigationTarget(site, HOME, ActionsState(tagsState?.tags)))
+            TAGS -> updateNavigationTarget(
+                    PrepublishingNavigationTarget(
+                            site,
+                            prepublishingScreen,
+                            tagsState
+                    )
+            )
+            PUBLISH -> TODO()
+            VISIBILITY -> TODO()
+        }
+    }
+
+    /**
+     * Checks to see if state was restored for a specific action that was ongoing. If not, then the Post Repository
+     * is used to create the state.
+     */
+    private fun updateScreenStatesFromPostRepository() {
+        if (tagsState == null) {
+            val tags = postRepository?.getPost()?.tagNameList
+            val formattedTags = tags?.let {
+                return@let formatTags(tags)
+            } ?: run {
+                throw IllegalArgumentException("Post or PostRepository can't be null.")
+            }
+            tagsState = TagsState(formattedTags)
+        }
+        // TODO add other states
+    }
+
+    /**
+     * Clicking back triggers the save operation.
+     */
+    fun onBackClicked() {
+        currentScreenState?.let { (screen, state) ->
+            when (screen) {
+                TAGS -> updateTagsAndState((state as? TagsState)?.tags)
+                HOME -> TODO()
+                PUBLISH -> TODO()
+                VISIBILITY -> TODO()
+            }
+            clearCurrentScreenState()
+            navigateToScreen(HOME)
+        } ?: run {
+            _dismissBottomSheet.postValue(Event(Unit))
+        }
+    }
+
+    private fun updateNavigationTarget(target: PrepublishingNavigationTarget) {
+        _navigationTarget.postValue(Event(target))
+    }
+
+    fun writeToBundle(outState: Bundle) {
+        outState.putParcelable(KEY_SCREEN_STATE, currentScreenState?.second)
+    }
+
+    fun onActionClicked(actionType: ActionType) {
+        val screen = PrepublishingScreen.valueOf(actionType.name)
+        clearCurrentStateAndTrackCurrentScreen(screen)
+        navigateToScreen(screen)
+    }
+
+    /**
+     * clear the current screen state since we are back at the home screen.
+     */
+    private fun clearCurrentScreenState() {
+        currentScreenState = null
+    }
+
+    private fun clearCurrentStateAndTrackCurrentScreen(prepublishingScreen: PrepublishingScreen) {
+        currentScreenState = Pair(prepublishingScreen, null)
+    }
+
+    fun updateTagsStateAndSetToCurrent(tags: String) {
+        tagsState = TagsState(tags)
+        currentScreenState = Pair(TAGS, tagsState)
+    }
+
+    /**
+     * This function updates the tags within the repository.
+     */
+    private fun updateTagsAndState(selectedTags: String?) {
```

Comment:
I'm wondering if this ViewModel doesn't too much. It seems it's basically a viewModel for all the subscreens + the main screen. 
Its current responsibilities are
1. Keeping state of all the subscreens
2. Storing and restoring the state of all the subscreens
3. Navigation between home and the subscreens
4. Specific actions of each subscreen such as `formatTags`

Have we considered creating either 
- a VM per subscreen
or
- a new class per subscreen which would be responsible for the tasks related to that subscreen

Don't change anything yet, this is just a discussion starter :).

Reply:
Thanks for this, yes it had way too many responsibilities. This ended up happening because of how intertwined the state persistence was with the `EditPostRepository`, `SavedInstanceState` and overall screens that needed the data. I have utilized this approach to clean up the code and move all the relevant responsibilities to use cases and viewmodels 🙏